lib/client/ipam_test.go

```diff
@@ -0,0 +1,300 @@
+// Test 3: claim affinity to a block twice.
+// - Claim affinity to an unclaimed block - expect to get the same IPNet back in claimed IP slice and no error.
+// - Claim affinity to the same block again - expect the IPNet back as the second return value (failed) and an error.
+
+package client_test
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"os/exec"
+	"strings"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	. "github.com/onsi/ginkgo/extensions/table"
+
+	"github.com/tigera/libcalico-go/calicoctl/commands"
+	"github.com/tigera/libcalico-go/lib/api"
+	"github.com/tigera/libcalico-go/lib/backend/etcd"
+	"github.com/tigera/libcalico-go/lib/client"
+	cnet "github.com/tigera/libcalico-go/lib/net"
+)
+
+// Setting BackendType to etcdv2 which is the only supported backend at the moment.
+var etcdType api.BackendType = "etcdv2"
+
+// Setting localhost as the etcd endpoint location since that's where `make run-etcd` runs it.
+var etcdConfig = etcd.EtcdConfig{
+	EtcdEndpoints: "http://127.0.0.1:2379",
+}
+
+var _ = Describe("IPAM tests", func() {
+
+	DescribeTable("AutoAssign: requested IPs vs returned IPs",
+		func(host string, cleanEnv bool, pool string, inv4, inv6, expv4, expv6 int, expError error) {
+			outv4, outv6, outError := testIPAMAutoAssign(inv4, inv6, host, cleanEnv, pool)
+			Expect(outv4).To(Equal(expv4))
+			Expect(outv6).To(Equal(expv6))
+			if expError != nil {
+				Ω(outError).Should(HaveOccurred())
+			}
+		},
+
+		// Test 1: AutoAssign 1 IPv4, 1 IPv6 - expect one of each to be returned.
+		Entry("1 v4 1 v6", "testHost", true, "pool1", 1, 1, 1, 1, nil),
+
+		// Test 2: AutoAssign 256 IPv4, 256 IPv6 - expect 256 IPv4 + IPv6 addresses.
+		Entry("256 v4 256 v6", "testHost", true, "pool1", 256, 256, 256, 256, nil),
+
+		// Test 3: AutoAssign 257 IPv4, 0 IPv6 - expect 256 IPv4 addresses, no IPv6, and no error.
+		Entry("257 v4 0 v6", "testHost", true, "pool1", 257, 0, 256, 0, nil),
+
+		// Test 4: AutoAssign 0 IPv4, 257 IPv6 - expect 256 IPv6 addresses, no IPv6, and no error.
+		Entry("0 v4 257 v6", "testHost", true, "pool1", 0, 257, 0, 256, nil),
+
+		// Test 5: (use pool of size /25 (/test/pool2.yaml) so only two blocks are contained):
+		// - Assign 1 address on host A (Expect 1 address).
+		Entry("1 v4 0 v6 host-A", "host-A", true, "pool2", 1, 0, 1, 0, nil),
+
+		// - Assign 1 address on host B (Expect 1 address, different block).
+		Entry("1 v4 0 v6 host-B", "host-B", false, "pool2", 1, 0, 1, 0, nil),
+
+		// - Assign 64 more addresses on host A (Expect 63 addresses from host A's block, 1 address from host B's block).
+		Entry("64 v4 0 v6 host-A", "host-A", false, "pool2", 64, 0, 64, 0, nil),
+	)
+
+	DescribeTable("AssignIP: requested IP vs returned error",
+		func(inIP net.IP, host string, cleanEnv bool, pool string, expError error) {
+			outError := testIPAMAssignIP(inIP, host, pool, cleanEnv)
+			if expError != nil {
+				Ω(outError).Should(HaveOccurred())
```

Comment:
You're right, I got confused by gomega documentation.
I fixed it. It's: `Expect(outError).To(HaveOccurred())`  now